gnomad/utils/constraint.py

```diff
@@ -877,3 +839,344 @@ def compute_expected_variants(
+    lower_idx_expr = hl.argmax(
+        norm_dpois_expr.map(lambda x: hl.or_missing(x < alpha, x))
+    )
+    upper_idx_expr = hl.argmin(
+        norm_dpois_expr.map(lambda x: hl.or_missing(x > 1 - alpha, x))
+    )
+    return hl.struct(
+        lower=hl.if_else(obs_expr > 0, range_expr[lower_idx_expr], 0),
+        upper=range_expr[upper_idx_expr],
+    )
+
+
+def calculate_raw_z_score(
+    obs_expr: hl.expr.Int32Expression,
+    exp_expr: hl.expr.Float32Expression,
+) -> hl.expr.StructExpression:
+    """
+    Compute the signed raw z-score using observed and expected variant counts.
+
+    The raw z-scores are positive when the transcript had fewer variants than expected,
+    and are negative when transcripts had more variants than expected.
+
+    :param obs_expr: Observed variant count expression.
+    :param exp_expr: Expected variant count expression.
+    :return: StrucExpression for the raw z-score.
+    """
+    chisq_expr = (obs_expr - exp_expr) ** 2 / exp_expr
+    return hl.sqrt(chisq_expr) * hl.if_else(obs_expr > exp_expr, -1, 1)
+
+
+def add_constraint_flags(
+    exp_expr: hl.expr.Float32Expression, outlier_expr: hl.expr.BooleanExpression
+) -> hl.expr.SetExpression:
+    """
+    Determine the constraint flags that define why constraint will not be calculated.
+
+    Flags which are added:
+        - "no_exp" - for genes that have zero expected variants
+        - "outlier" - for genes that are outliers according to the supplied `outlier_expr`
+
+     :param exp_expr: Expression for the expected variant counts of pLoF, missense, or
+        synonymous variants.
+     :param outlier_expr: Boolean expression used to decide if an 'outlier' flag should
+        be applied (example: ht['lof'].z_raw < -5).
+     :return: SetExpression containing constraint flags.
+    """
+    constraint_flags = hl.empty_set(hl.tstr)
+    constraint_flags = hl.if_else(
+        exp_expr <= 0,
+        constraint_flags.add(f"no_exp"),
+        constraint_flags,
+    )
+
+    constraint_flags = hl.if_else(
+        outlier_expr,
+        constraint_flags.add(f"outlier"),
+        constraint_flags,
+    )
+
+    return constraint_flags
+
+
+def calculate_z_score(
+    ht: hl.Table,
+    raw_z_expr: hl.expr.StringExpression,
+    flag_expr: hl.expr.StringExpression,
+    additional_requirements_expr: hl.expr.BooleanExpression,
+    both: bool = True,
+) -> hl.expr.StructExpression:
+    """
+    Calculate the standard deviation of the raw z-Score and the z-score.
+
+    :param ht: Input Table with raw z-score and constraint flag expression.
+    :param raw_z_expr: Expression for the raw z-score.
+    :param flag_expr: Expression for the constraint flags. z-score will not be
+        calculated if flags are present.
+    :param additional_requirements_expr: Expression for additional requirements that
+        must be met for z-scores to be calculated.
+    :param both: Whether to use both the positive and negative `raw_z_expr` when
```

Comment:
I added another parameter `neg_raw_z_only` to handle the differences with lof/mis and syn. I think I made it confusing though with the `both` parameter since they do different things, but sound similar in description. Thoughts on another name other than `both`?